planner/core/rule_join_elimination.go

```diff
@@ -0,0 +1,165 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"github.com/pingcap/parser/ast"
+	"github.com/pingcap/tidb/expression"
+)
+
+type outerJoinEliminator struct {
+	cols    [][]*expression.Column
+	schemas []*expression.Schema
+}
+
+// tryToEliminateOuterJoin will eliminate outer join plan base on the following rules
+// 1. outer join elimination: For example left outer join, if the parent only use the
+//    columns from left table and the join key of right table(the inner table) is a unique
+//    key of the right table. the left outer join can be eliminated.
+// 2. outer join elimination with distinct: For example left outer join. If the parent
+//    only use the columns from left table with 'distinct' label. The left outer join can
+//    be eliminated.
+func (o *outerJoinEliminator) tryToEliminateOuterJoin(p *LogicalJoin) LogicalPlan {
+	switch p.JoinType {
+	case LeftOuterJoin:
+		return o.doEliminate(p, 1)
+	case RightOuterJoin:
+		return o.doEliminate(p, 0)
+	default:
+		return nil
+	}
+}
+
+func (o *outerJoinEliminator) doEliminate(p *LogicalJoin, innerChildIdx int) LogicalPlan {
+	// outer join elimination with distinct
+	if len(o.cols) > 0 {
+		cols := o.cols[len(o.cols)-1]
+		allColsInSchema := true
+		for _, col := range cols {
+			columnName := &ast.ColumnName{Schema: col.DBName, Table: col.TblName, Name: col.ColName}
+			if c, _ := p.children[1^innerChildIdx].Schema().FindColumn(columnName); c == nil {
+				allColsInSchema = false
+				break
+			}
+		}
+		if allColsInSchema == true {
+			return p.children[1^innerChildIdx]
+		}
+	}
+
+	// outer join elimination without distinct
+	// first, check whether the parent's schema columns are all in left or right
+	if len(o.schemas) == 0 {
+		return nil
+	}
+	for _, col := range o.schemas[len(o.schemas)-1].Columns {
```

Comment:
It can be promised that `o.schemas[len(o.schemas)-1]` is always the schema of current Join?

Reply:
Yes. It's obviouse at line147~line151.